### PR TITLE
feat: add support for ERC-3770 on recipient addr

### PR DESCRIPTION
**Pull request by compojoom**
# Summary

Currently if you paste an ERC-3770 formatted address into the recipient list, you’ll get an error and have to delete the chain prefix.

With this change we will detect a prefix in an address and return just the address. If the address pasted was having the same address as the current chain, there won’t be any warning to the user, but if he has pasted an address for a different chain we’ll output a warning.

grafik

This improves the compatibility with the Safe app as addresses there are by default output with a prefix. 

# To Test

1. Enable custom recipient
2. If you enter an address with a prefix, then the prefix will be stripped out and only the address will be left in the field.
- if the prefix was for the same chain the user is currently on - no warning will be output
- if the prefix was for a different chain, then a warning will be output

3. If you paste an address without prefix, then it should continue to work as before.

### CODE REVIEW

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (anxolin) [inline apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:144]: What about using an object the type:

`const [chainPrefixWarning, setChainPrefixWarning] = useState<ChainPrefixWatning | undefined>(false)`

Where `ChainPrefixWatning` is:

```
interface ChainPrefixWatning {
   walletChainId: SupportedChain
   recipientChainId: SupportedChain
}
```

This way you can make a nicer warning message

Comment (anxolin) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:144]: or maybe just save the recipientChainId, as the walletChainId is known

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (anxolin) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:176]: I think we should review this message, and ideally make it closable. 

Also, what about changing the message to something in this lines: 
`The recipient address you inputted belongs to {recipientChainName} but your wallet is {walletChainName}. Please, make they are on the same network.`

**Review by anxolin — COMMENTED** [review]
Looks great! good job.

I added some small nitpick comments. Other than that, @elena-zh will need to test it. So for now, you have a soft approve.

I couldn't test it right now (placing an order, because staging API is returning some errors on the quotes), but apparently it works!

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (compojoom) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:132]: Hm, I did `yarn lint --fix` - no idea why my editor added them.

Comment (compojoom) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:132]: ok, it turns out that prettier was disabled in my editor... Will, reformat.

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (compojoom) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:88]: I got rid of that function and moved the logic to the getChecksumAddressOrOriginal.

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (compojoom) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx]: Check it out now.

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (compojoom) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx]: check it out now.

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (shoom3301) [inline apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:12]: Please, use `import { getAddress } from '@ethersproject/address'` instead

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (shoom3301) [inline apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:99]:
```suggestion
export const parsePrefixedAddress = (value: string) => {
  const [prefix, address] = value.split(':')

  return {
    prefix: prefix || undefined,
    address: address ? getChecksumAddressOrOriginal(address) : value,
  }
}
```

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (shoom3301) [inline apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:85]: Please, move utils to a separate file or even better to `libs/common-utils`

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (shoom3301) [inline apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:134]: The commented code should be deleted

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (shoom3301) [inline apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:166]: Would be great to move the component to `common/pure`

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (compojoom) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx:12]: Ok. When are you guys moving to ethers v6?

**Review by elena-zh — COMMENTED** [review]
Hey @compojoom , great job!
Some issues/nitpicks I faced: 
1. If a warning is displayed on a swap container, it would be great to show it on an order confirm modal as well. @anxolin , WDYT? 
conf

2. I noticed that sometimes the app shows a warning after a network change. To reproduce: 
- clear out your local storage
- open TWAP form on Ethereum
- Enable a Custom recipient
- change a chain using network selector to Sepolia
- paste sep:0x9FA3c00a92Ec5f96B1Ad2527ab41B3932EFEDa58 into the recipient field --> Warning will appear
image


4. I think, it would be great to track whether an address is a SC or EOA wallet. Here is the case: when I paste sep:0xFF714b8b0e2700303eC912BD40496C3997ceEa2b (which is EOA address) when connected to Ethereum network, it shows me a warning that makes me think that I've done something wrong. However, eventually there is nothing wrong as the funds will be received by EOA address in the selected network. 
eoa
3. (continuation): if a smart contract wallet is filled in into the Recipient, and it does not contain a chain indicator, i think it would be great to show a message something like 'Please double-check that this smart contract address exists in the selected chain.' or so :)

**Comment by elena-zh** [comment]
@compojoom , checked in the code, the new message looks the way better!
However, still it would be great to take a look at cases 1-2.

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (anxolin) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx]: Good question! we have a dependency with the contracts NPM package, so either when we migrate that or when we kill that dependency. Not urging for now.

Alternatively never, if we move to viem :P

**Review by anxolin — APPROVED** [review]
That's annoying that we need to approve every time we want to redeploy. 

Thanks for addressing the comments. 

Giving an approve on the code, leaving up to @elena-zh the final approval

**Comment by compojoom** [comment]
Hey @elena-zh, 

I gave it a bit more thought and here is how I see it:

> If a warning is displayed on a swap container, it would be great to show it on an order confirm modal as well. @anxolin , WDYT?

This warning is only displayed if the user pastes an EIP3770 address. (I don't think a user would ever manually type an EIP3770 address, but if he does we'll strip it from the prefix as well :)). The warning is immediate & I think it is sufficient to draw attention. Actually my main goal was not to disturb the user, but to actually make it possible to paste an EIP-3770 address. This way the user won't have to go and delete the prefix. The warning in my view is just the cherry on top :)

> I noticed that sometimes the app shows a warning after a network change

I guess that it would be possible to reset the warning on switching a network, but maybe it is not bad to have it?

> I think, it would be great to track whether an address is a SC or EOA wallet. Here is the case: when I paste sep:0xFF714b8b0e2700303eC912BD40496C3997ceEa2b (which is EOA address) when connected to Ethereum network, it shows me a warning that makes me think that I've done something wrong. However, eventually there is nothing wrong as the funds will be received by EOA address in the selected network.

This is tricky & too much for this PR. Like I like to tell my AI "let's think this through, step by step" :)
User is on ETH. Pastes a bnb address. We would have to check the bnb chain to see if there was a deployed contract to know if this is a contract address, then check eth if there is contract for this address and then warn. This would mean that we would have to configure a custom RPC for every possible chain in order to check whether the address pasted is a contract or not. And even if it isn't, there is no guarantee that it won't become one.

> (continuation): if a smart contract wallet is filled in into the Recipient, and it does not contain a chain indicator, i think it would be great to show a message something like 'Please double-check that this smart contract address exists in the selected chain.' or so :)

How do we know if it is a smart contract wallet? We cannot determine this based on the address and would need to perform a network request to see if there is a contract deployed at this address. And if there is a contract, then it is not necessarily a smart contract wallet. It could be any kind of contract. 

Those heuristics are never accurate and for now I think that we have to rely on a human's ability to think and be vigilant.

**Comment by elena-zh** [comment]
Hey @compojoom , thank you for your answers. I agree with you on cases 1, 3, 4. These are nitpicks and require a lot of effort for implementation. 
But I can't agree with this:

>  > I noticed that sometimes the app shows a warning after a network change
> 
> I guess that it would be possible to reset the warning on switching a network, but maybe it is not bad to have it?

For me it looks weird to see this message when I'm on Gnosis chain, and I've pasted an address with GNO in the prefix:

 _The recipient address you inputted had the chainPrefix **gno**, but your wallet is on **Gnosis chain** which uses the chain prefix **gno**. Please, double-check that the recipient address is for the correct network._

**Comment by anxolin** [comment]
@elena-zh but the warning is shown only if someone paste an address with the prefix and then change network, right? 

Practically speaking this will only happen in the Safe, where it doesn't make sense to change the network. Each safe is bound to a network. 

From my side, although desirable I wouldn't block merging this feature for these

**Comment by elena-zh** [comment]
> @elena-zh but the warning is shown only if someone paste an address with the prefix and then change network, right?
No. The warning is displayed when I 
1. change a network
2. then paste an address with the **correct** prefix. 

I described steps [here ](https://github.com/cowprotocol/cowswap/pull/4418#pullrequestreview-2057914197)
image

**Comment by compojoom** [comment]
hey @elena-zh, 
For some reason I could not reproduce it, but I have a hunch, that the problem lies in callback dependencies. I've added the chainInfo.prefix which should change when the user switch the chain to it. Can you try again? Please first trigger a rebuild :)

**Review by shoom3301 — APPROVED** [review]

File: apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx
Comment (jxom) [reply apps/cowswap-frontend/src/legacy/components/AddressInputPanel/index.tsx]: happy to help if migration to viem is considered! :D

**Comment by elena-zh** [comment]
Hey @compojoom , thank you so much! Now it works for me. 
One last nitpick: it would be great to write 'chain prefix' in the same manner in the sentence
chain prefix
Also, the message is may be displayed when a user is not connected to a wallet. so WDYT about a bit more generic 
'`The recipient address you inputted had the chain prefix gno, but you are on Sepolia which uses the chain prefix sep. Please, double-check that the recipient address is for the correct network`.'?

**Comment by anxolin** [comment]
@elena-zh I will merge, lets iterate ourselves to keep this moving.

Thanks @compojoom for your first contribution to CoW Swap! Congrats! 🥂

**Comment by anxolin** [comment]
Merged, and changed the message slightly https://github.com/cowprotocol/cowswap/pull/4468

**Pull request merged**
